Mathlib/CategoryTheory/Limits/Opposites.lean

```diff
@@ -414,44 +414,64 @@ instance : HasProduct (fun z ↦ op (Z z)) := hasLimitOfIso
     Discrete.functor (fun z ↦ op (Z z)))
 
 /-- The isomorphism from the opposite of the coproduct to the product. -/
-@[simp]
 noncomputable
 def opCoproductIsoProduct : op (∐ Z) ≅ ∏ (fun z => op (Z z)) :=
   IsLimit.conePointUniqueUpToIso (isLimitCoconeOp _ (coproductIsCoproduct fun b ↦ Z b))
     (limit.isLimit _) ≪≫ (IsLimit.conePointsIsoOfEquivalence
     (productIsProduct (fun z ↦ op (Z z))) (limit.isLimit _) (Discrete.opposite α).symm
     (Discrete.natIsoFunctor ≪≫ Discrete.natIso (fun _ ↦ by rfl))).symm
 
-lemma opCoproductIsoProduct_inv_comp_ι (b : α) :
-    (opCoproductIsoProduct Z).inv ≫ (Sigma.ι (fun a => Z a) b).op =
+lemma opCoproductIsoProduct_inv_comp_ι' {c : Cofan Z} (hc : IsColimit c) (b : α) :
+    (opCoproductIsoProduct Z).inv ≫
```

Comment:
This is a good effort! But I think we should push this one last step further by refactoring the definition of `opCoproductIsoProduct` as `opCoproductIsoProduct'` which would take as inputs a colimit cofan and a limit fan fan in the opposite category, like:
```lean

variable {C : Type*} [Category C] {α : Type*} {Z : α → C}

def Cofan.op {c : Cofan Z} : Fan (fun a ↦ op (Z a)) := Fan.mk _ (fun a => (c.inj a).op)

def Cofan.IsColimit.op {c : Cofan Z} (hc : IsColimit c) : IsLimit c.op := by
  let e : Discrete.functor (fun a => Opposite.op (Z a)) ≅ (Discrete.opposite α).inverse ⋙ (Discrete.functor Z).op :=
    Discrete.natIso (fun _ => Iso.refl _)
  refine' IsLimit.ofIsoLimit ((IsLimit.postcomposeInvEquiv e _).2
    (IsLimit.whiskerEquivalence hc.op (Discrete.opposite α).symm))
    (Cones.ext (Iso.refl _) (fun ⟨a⟩ => _))
  dsimp
  erw [Category.id_comp, Category.comp_id]
  rfl

def opCoproductIsoProduct' {c : Cofan Z} {f : Fan (fun a ↦ op (Z a))}
    (hc : IsColimit c) (hf : IsLimit f) : op c.pt ≅ f.pt :=
  IsLimit.conePointUniqueUpToIso (Cofan.IsColimit.op hc) hf

def opCoproductIsoProduct :
    op (∐ Z) ≅ ∏ (fun z => op (Z z)) :=
  opCoproductIsoProduct' (coproductIsCoproduct Z) (productIsProduct (fun z ↦ op (Z z)))
```
This should also simplify the proofs of equational lemmas for `opCoproductIsoProduct` and `opCoproductIsoProduct'`.

Reply:
Note that it is probably better to do this in a separate PR.

Reply:
@joelriou #8302 (I did the dual as well)